app/src/lib/git/rebase.ts

```diff
@@ -0,0 +1,159 @@
+import * as Path from 'path'
+import * as FSE from 'fs-extra'
+
+import { Repository } from '../../models/repository'
+import { git } from './core'
+import {
+  WorkingDirectoryFileChange,
+  AppFileStatusKind,
+} from '../../models/status'
+import { ManualConflictResolution } from '../../models/manual-conflict-resolution'
+import { stageManualConflictResolution } from './stage'
+import { stageFiles } from './update-index'
+import { GitError, IGitResult } from 'dugite'
+import { getStatus } from './status'
+
+/**
+ * Check the `.git/REBASE_HEAD` file exists in a repository to confirm
+ * a rebase operation is underway.
+ */
+export function isRebaseHeadSet(repository: Repository) {
+  const path = Path.join(repository.path, '.git', 'REBASE_HEAD')
+  return FSE.pathExists(path)
+}
+
+/**
+ * A stub function to use for initiating rebase in the app.
+ *
+ * If the rebase fails, the repository will be in an indeterminate state where
+ * the rebase is stuck.
+ *
+ * If the rebase completes without error, `featureBranch` will be checked out
+ * and it will probably have a different commit history.
+ *
+ * @param baseBranch the ref to start the rebase from
+ * @param featureBranch the ref to rebase onto `baseBranch`
+ */
+export async function rebase(
+  repository: Repository,
+  baseBranch: string,
+  featureBranch: string
+) {
+  return await git(
+    ['rebase', baseBranch, featureBranch],
+    repository.path,
+    'abortRebase',
+    { expectedErrors: new Set([GitError.RebaseConflicts]) }
+  )
+}
+
+/** Abandon the current rebase operation */
+export async function abortRebase(repository: Repository) {
+  await git(['rebase', '--abort'], repository.path, 'abortRebase')
+}
+
+export enum ContinueRebaseResult {
+  CompletedWithoutError = 'CompletedWithoutError',
+  ConflictsEncountered = 'ConflictsEncountered',
+  OutstandingFilesNotStaged = 'OutstandingFilesNotStaged',
+  Aborted = 'Aborted',
+}
+
+const rebaseEncounteredConflictsRe = /Resolve all conflicts manually, mark them as resolved/
+
+const filesNotMergedRe = /You must edit all merge conflicts and then\smark them as resolved/
```

Comment:
Replacing `\s` with literal whitespace would make this slightly easier to read and more consistent with the rest of the expression and with the one above.

Reply:
I want to leave this as `\s` because I was seeing this output break over a newline, but maybe I should be explicit here and use `\n` instead. Thinking about it some more now 🤔

Reply:
Yep, a space here fails the tests (note the `\n` usages):

```
Unhandled result found: '{"stdout":"OTHER.md: needs merge\nTHING.md: needs merge\nYou must edit all merge conflicts and then\nmark them as resolved using git add\n","stderr":"","exitCode":1,"gitError":null,"gitErrorDescription":null}'
```

Switched to `\n` in 51fee1742